src/Limiter.h

```diff
@@ -113,6 +115,51 @@ class Limiter : public ProcessPlugin
   int getNumOutputs() override { return(mNumChannels); }
   void compute(int nframes, float** inputs, float** outputs) override;
 
+  void setWarningAmplitude(double wa) { // setting to 0 turns off warnings
+    warningAmp = std::max(0.0,std::min(1.0,wa));
+  }
+
+ private:
+
+  void checkAmplitudes(int nframes, float* buf) {
+    const int maxWarningInterval { 10000 }; // this could become an option
+    assert(warningAmp > 0.0);
+    assert(mNumClients > 0);
+    for (int i=0; i<nframes; i++) {
+      double tmp_sample = double(buf[i]);
+      double limiterAmp = fabs(tmp_sample)/sqrt(double(mNumClients)); // KEEP IN SYNC with gain in ../faust-src/limiterdsp.dsp
+      if (limiterAmp >= warningAmp) {
+        warnCount++;
+        peakMagnitude = std::max(peakMagnitude,limiterAmp);
+        if (warnCount==nextWarning) {
+          double peakMagnitudeDB = 20.0 * std::log10(peakMagnitude);
+          double warningAmpDB = 20.0 * std::log10(warningAmp);
+          if (warnCount==1) {
+            if (warningAmp == 1.0) {
+              std::cerr << "*** Limiter.cpp: Audio HARD-CLIPPED!\n";
+              fprintf(stderr, "\tReduce your audio input level(s) by %0.1f dB to avoid this.\n", peakMagnitudeDB);
+            } else {
+              fprintf(stderr,
+                      "*** Limiter.cpp: Amplitude levels must stay below %0.1f dBFS to avoid compression.\n",
+                      warningAmpDB);
+              fprintf(stderr, "\tReduce input level(s) by %0.1f dB to achieve this.\n",
+                      peakMagnitudeDB-warningAmpDB);
+            }
+          } else {
+            fprintf(stderr, "\tReduce audio input level(s) by %0.1f dB to avoid limiter compression distortion.\n",
+                    peakMagnitudeDB-warningAmpDB);
+          }
+          peakMagnitude = 0.0; // reset for next group measurement
+          if (nextWarning < maxWarningInterval) { // don't let it stop reporting for too long
+            nextWarning *= 10;
+          } else {
+            warnCount=0;
+          }
+        } // warnCount==nextWarning
```

Comment:
This code looks familiar.. is there a way we could reuse it versus having two copies?

Reply:
There is actually only one copy, because the copy in AudioInterface.cpp is #ifdef'd out.  I am expecting Chris to choose to delete the more general warning in AudioInterface.cpp, in which case the apparent 2nd copy will disappear.  On the other hand, if we do want such a feature, then we can make a shared utility.  However, I see no demand for that.

Reply:
To simplify looking at this, I removed the disabled code blocks, but kept them in the jos branch, both to facilitate reviving them if anyone wants them, and to give me a double-check that the warnings in AudioInterface.cpp never fire regardless of which limiter I'm using.